### PR TITLE
Add databases support

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

**Comment by tsmr** [comment]
Hello : I see this PR :)

glpi_plugin_databases_databases_items and glpi_plugin_databases_scripts tables are missing

It's normal ?

**Comment by trasher** [comment]
Hi,

Nothing is really missing since I'm completely writting that from scratch; the main goal is to add the support for some kind of "inventory" of databases. 
Also keep in mind this is currently a draft and still requires a lot of work ;)

Anyways could you explain what's the purpose of the tables you're talking about? Thanks!

**Comment by tsmr** [comment]
glpi_plugin_databases_databases_items : for linking database to computer etc..

glpi_plugin_databases_scripts : for adding scripts linked to a database (script on database launch..)

File: inc/database.class.php [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+   }
+
+
+   function post_addItem() {
+      //create default instance
+      $this->getFromDB($this->fields['id']);
+      $instance = new DatabaseInstance();
+      $instance->add([
+         'name' => ($this->input['instance_name'] ?? $this->input['name']),
+         'port' => ($this->input['port'] ?? 0),
```
Comment (cedric-anne) [inline inc/database.class.php:248]:
```suggestion
         'port' => ($this->input['instance_port'] ?? 0),
```

File: inc/database.class.php [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+         'size' => ($this->input['size'] ?? 0),
+         'is_onbackup' => ($this->input['is_onbackup'] ?? 0),
```
Comment (cedric-anne) [inline inc/database.class.php:250]:
```suggestion
         'is_onbackup' => ($this->input['instance_is_onbackup'] ?? 0),
```

File: inc/database.class.php [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+         'is_active' => ($this->input['instance_is_active'] ?? $this->fields['is_active'] ?? 0),
```
Comment (cedric-anne) [inline inc/database.class.php:251]: I guess `$this->fields['is_active']` is always defined.
```suggestion
         'is_active' => ($this->input['instance_is_active'] ?? $this->fields['is_active']),
```

File: inc/database_item.class.php [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+         return '';
+      }
+
+      $nb = 0;
+      if ($item->getType() == Database::class) {
+         if ($_SESSION['glpishow_count_on_tabs']) {
+            if (!$item->isNewItem()) {
+               $nb = self::countForMainItem($item);
+            }
+         }
```
Comment (cedric-anne) [inline inc/database_item.class.php:64]:
```suggestion
         if ($_SESSION['glpishow_count_on_tabs'] && !$item->isNewItem()) {
            $nb = self::countForMainItem($item);
         }
```

File: inc/database_item.class.php [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+         default :
+            if (in_array($item->getType(), Database::getTypes())) {
+               self::showForItem($item, $withtemplate);
+            }
+      }
+      return true;
+   }
+
+   /**
+    * Print enclosure items
```
Comment (cedric-anne) [inline inc/database_item.class.php:90]:
```suggestion
    * Print database items
```

File: inc/database_item.class.php [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+            $header .= Html::getCheckAllAsCheckbox('mass'.__CLASS__.$rand);
+            $header .= "</th>";
+         }
+         $header .= "<th>".__('Itemtype')."</th>";
+         $header .= "<th>"._n('Item', 'Items', 1)."</th>";
+         $header .= "</tr>";
+         echo $header;
+
+         foreach ($items as $row) {
+            $item = new $row['itemtype'];
```
Comment (cedric-anne) [inline inc/database_item.class.php:184]: Would be safer to check itemtype validity.
```suggestion
            if (!($item = getItemForItemtype($row['itemtype']))) {
               continue;
            }
```

File: inc/database_item.class.php [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+               if ($_SESSION["glpiis_ids_visible"]
+                  || empty($database->fields["name"])) {
+                  $name = sprintf(__('%1$s (%2$s)'), $name, $database->fields["id"]);
+               }
+               echo "<a href='" . Database::getFormURLWithID($cID) . "'>" . $name . "</a>";
+               echo "<td>".$instance->fields['name']."</td>";
+               echo "<td>".$instance->fields['port']."</td>";
+               echo "<td>".$instance->fields['size']."</td>";
+               echo "<td>".Dropdown::getYesNo($instance->fields['is_active'])."</td>";
+               echo "</td>";
```
Comment (cedric-anne) [inline inc/database_item.class.php:326]:
```suggestion
```

File: inc/database_item.class.php [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+                  Html::showMassiveActionCheckBox(__CLASS__, $assocID);
+                  echo "</td>";
+               }
+               echo "<td class='b'>";
+               $name = $database->fields["name"];
+               if ($_SESSION["glpiis_ids_visible"]
+                  || empty($database->fields["name"])) {
+                  $name = sprintf(__('%1$s (%2$s)'), $name, $database->fields["id"]);
+               }
+               echo "<a href='" . Database::getFormURLWithID($cID) . "'>" . $name . "</a>";
```
Comment (cedric-anne) [inline inc/database_item.class.php:321]:
```suggestion
               echo "<a href='" . Database::getFormURLWithID($cID) . "'>" . $name . "</a>";
               echo "</td>";
```

File: inc/database_item.class.php [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+         }
+         return false;
+      }
+
+      return $input;
+   }
+
+   public static function countForMainItem(CommonDBTM $item, $extra_types_where = []) {
+      $types = [];//Database::getTypes();
+      Appliance::getTypes();
```
Comment (cedric-anne) [inline inc/database_item.class.php:395]:
```suggestion
      $types = Database::getTypes();
```

File: inc/define.php [hunk 10]
```diff
@@ -571,6 +579,7 @@
    Printer::getType()            => "pics/impact/printer.png",
    Rack::getType()               => "pics/impact/rack.png",
    Software::getType()           => "pics/impact/software.png",
+   Database::getType()           => "",
```
Comment (cedric-anne) [inline inc/define.php:582]: TODO

File: install/migrations/update_9.5.x_to_10.0.0/databases.php [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+         `databasetypes_id` int NOT NULL DEFAULT '0',
+         `databasecategories_id` int NOT NULL DEFAULT '0',
+         `locations_id` int NOT NULL DEFAULT '0',
+         `manufacturers_id` int NOT NULL DEFAULT '0',
+         `suppliers_id` int NOT NULL DEFAULT '0',
+         `users_id_tech` int NOT NULL DEFAULT '0',
+         `groups_id_tech` int NOT NULL DEFAULT '0',
+         `date_mod` timestamp NULL DEFAULT NULL,
+         `date_creation` timestamp NULL DEFAULT NULL,
+         `date_lastboot` timestamp NULL DEFAULT NULL,
```
Comment (cedric-anne) [inline install/migrations/update_9.5.x_to_10.0.0/databases.php:89]: I guess a database has always an instance. Is this `date_lastboot` is relevant if we already have one for each instances ?

File: inc/database.class.php [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+
+      $this->showFormButtons($options);
+
+      return true;
+   }
+
+   function rawSearchOptions() {
+
+      $tab = parent::rawSearchOptions();
+
```
Comment (cedric-anne) [inline inc/database.class.php:235]: Some more search options could be usefull.

File: inc/databaseinstance.class.php [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+            Session::addToNavigateListItems('DatabaseInstance', $instance['id']);
+
+         }
+         echo $header;
+      }
+      echo "</table>";
+      echo "</div>";
+   }
+
+   function showForm($ID, $options = []) {
```
Comment (cedric-anne) [inline inc/databaseinstance.class.php:157]: A link to parent database could be usefull in this form.

File: front/databaseinstance.form.php [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+      Event::log($_POST['databases_id'], "databases", 4, "management",
+                 //TRANS: %s is the user login
+                 sprintf(__('%s adds a database instance'), $_SESSION["glpiname"]));
+      if ($_SESSION['glpibackcreated']) {
+         Html::redirect($instance->getLinkURL());
+      }
+   }
+   Html::back();
+
+} else if (isset($_POST["purge"])) {
```
Comment (cedric-anne) [inline front/databaseinstance.form.php:61]: `$_POST["delete"]` and `$_POST["delete"]` cases are missing

File: inc/databaseinstance.class.php [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+         echo "</a></div>\n";
+      }
+
+      echo "<div class='center'>";
+
+      $instances = getAllDataFromTable(
+         self::getTable(), [
+            'WHERE'  => [
+               'databases_id' => $ID,
+               'is_deleted'   => 0
```
Comment (cedric-anne) [inline inc/databaseinstance.class.php:113]: Maybe deleted instances should be displayed, in order to be able to restore them.

File: front/databaseinstance.form.php [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+   $_GET["id"] = "";
+}
+
+if (!isset($_GET["databases_id"])) {
+   $_GET["databases_id"] = "";
+}
+
+$instance = new DatabaseInstance();
+if (isset($_POST["add"])) {
+   $instance->check(-1, CREATE, $_POST);
```
Comment (cedric-anne) [inline front/databaseinstance.form.php:49]: I get a `You don't have permission to perform this action.` when trying to add a new DB instance.

Comment (trasher) [reply front/databaseinstance.form.php:49]: Strange... Maybe you/I forget to disconnect; I'll double check

File: inc/database.class.php [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+            'instance_port', [
+               'type' => 'number',
+               'id' => 'instance_port'.$rand
+            ]
+         );
+
+         echo "</td>";
+         echo "<td><label for='dropdown_is_onbackup$rand'>".__('Has backup')."</label></td>";
+         echo "<td>";
+         Dropdown::showYesNo('is_onbackup', 0, -1, ['rand' => $rand]);
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/database.class.php:221]: That change requires some others; but OK, I'll do it

File: inc/database.class.php (hunk 12 above)
Comment (trasher) [reply inc/database.class.php:235]: Sure, sounds like I forget that :)

File: inc/define.php (hunk 10 above)
Comment (trasher) [reply inc/define.php:582]: Yes, but I do not have any idea where to find an image.

File: inc/database.class.php [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+            'maybeempty' => true
+         ]);
+      echo "</td></tr>\n";
+
+      echo "<tr class='tab_bg_1'>";
+      echo "<td><label for='is_active$rand'>".__('Is active')."</label></td>";
+      echo "<td>";
+      Dropdown::showYesNo('is_active', $this->fields['is_active']);
+      echo "<td>" . __('Associable to a ticket') . "</td><td>";
+      Dropdown::showYesNo('is_helpdesk_visible', $this->fields['is_helpdesk_visible']);
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/database.class.php:165]: I really do not know; I took Appliance as model for this class. ping @orthagh

File: install/migrations/update_9.5.x_to_10.0.0/databases.php (hunk 11 above)
Comment (trasher) [reply install/migrations/update_9.5.x_to_10.0.0/databases.php:89]: In original specifications; last boot date is set only on databases; and not at all on instances; but maybe should it be set only on instances indeed. Any thoughts @orthagh @stonebuzz?

**Comment by trasher** [comment]
> glpi_plugin_databases_databases_items : for linking database to computer etc..

OK, it's been added.

> glpi_plugin_databases_scripts : for adding scripts linked to a database (script on database launch..)

Well, this is not planned to be added on my side; but I'm not against it.

File: install/migrations/update_9.5.x_to_10.0.0/databases.php (hunk 11 above)
Comment (trasher) [reply install/migrations/update_9.5.x_to_10.0.0/databases.php]: I've dropped that lastboot field

File: front/databaseinstance.form.php (hunk 16 above)
Comment (cedric-anne) [reply front/databaseinstance.form.php:49]: Indeed, it is OK.

File: inc/databaseinstance.class.php [hunk 19]
```diff
@@ -0,0 +1,349 @@
...
+      echo "<td><label for='port$rand'>"._n('Port', 'Ports', 1)."</label></td>";
+      echo "<td>";
+      echo Html::input(
+         'port', [
+            'id' => 'port'.$rand,
+            'value' => $this->fields['port']
+         ]
+      );
+      echo "</td>";
+      echo "<td><label for='size$rand'>".__('Size')."</label></td>";
```
Comment (cedric-anne) [inline inc/databaseinstance.class.php:192]: Should size unit be mentionned in label ?

File: inc/databaseinstance.class.php [hunk 20]
```diff
@@ -0,0 +1,349 @@
...
+ */
+
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access this file directly");
+}
+
+class DatabaseInstance extends CommonDBChild {
+
+   // From CommonDBTM
+   public $auto_message_on_action   = false;
```
Comment (cedric-anne) [inline inc/databaseinstance.class.php:40]: When a database instance is added and when `$_SESSION['glpibackcreated']` is not activated, user is redirected to an empty creation form without even a message that indicates that the instance has been created.
Following change fixes this.

```suggestion
   public $auto_message_on_action   = true;
```

File: inc/databaseinstance.class.php [hunk 21]
```diff
@@ -0,0 +1,349 @@
...
+      echo "<td>";
+      Dropdown::showYesNo('is_active', $this->fields['is_active']);
+      echo "</td></tr>\n";
+
+      echo "<tr class='tab_bg_1'>";
+      echo "<td><label for='port$rand'>"._n('Port', 'Ports', 1)."</label></td>";
+      echo "<td>";
+      echo Html::input(
+         'port', [
+            'id' => 'port'.$rand,
```
Comment (cedric-anne) [inline inc/databaseinstance.class.php:187]:
```suggestion
            'id' => 'port'.$rand,
            'type' => 'number',
```

File: inc/databaseinstance.class.php [hunk 22]
```diff
@@ -0,0 +1,349 @@
...
+            'value' => $this->fields['port']
+         ]
+      );
+      echo "</td>";
+      echo "<td><label for='size$rand'>".__('Size')."</label></td>";
+      echo "<td>";
+      echo Html::input(
+         'size', [
+            'id' => 'size'.$rand,
```
Comment (cedric-anne) [inline inc/databaseinstance.class.php:196]:
```suggestion
            'id' => 'size'.$rand,
            'type' => 'number',
```

File: inc/database.class.php [hunk 23]
```diff
@@ -0,0 +1,415 @@
...
+
+      if ((!isset($this->input['instance_name']) || empty($this->input['instance_name'])) && isset($this->input['name'])) {
+         $this->input['instance_name'] = sprintf(
+            //TRANS: first parameter is the database name
+            __('"%1$s" default instance'),
+            $this->input['name']
+         );
+      }
+
+      $instance->add([
```
Comment (cedric-anne) [inline inc/database.class.php:349]: I wonder if adding a default instance when all values are empty is really usefull.

File: inc/inventory/asset/database.class.php [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+      foreach ($db_instances as $keydb => $arraydb) {
+         //update existing databases
+         $dbid = $arraydb['dbid'];
+         unset($arraydb['dbid']);
+         //TODO: update database, its instance, and link to item
+         foreach ($value as $key => $val) {
+            $sinput = [
+               'name' => $val->name ?? '',
+               'instance_name' => $val->instance_name ?? '',
+               'port' => $val->port ?? '',
```
Comment (cedric-anne) [inline inc/inventory/asset/database.class.php:124]:
```suggestion
               'instance_port' => $val->port ?? '',
```

File: inc/inventory/asset/database.class.php [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+      $database = new GDatabase();
+      $dbitem = new \Database_Item();
+
+      $db_instances = $this->getExisting();
+
+      foreach ($db_instances as $keydb => $arraydb) {
+         //update existing databases
+         $dbid = $arraydb['dbid'];
+         unset($arraydb['dbid']);
+         //TODO: update database, its instance, and link to item
```
Comment (cedric-anne) [inline inc/inventory/asset/database.class.php:119]: TODO

File: inc/database.class.php (hunk 23 above)
Comment (trasher) [reply inc/database.class.php:343]: Well, it's been discussed; and the point was there is always an instance with a database; whether we got its information or not.
The API should support adding instances information; and there are a few information (like port) that can be specified when creating a database from the UI.

File: inc/inventory/asset/database.class.php [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+
+      $db_existing = [];
+
+      $iterator = $DB->request([
+         'SELECT' => [
+            GDatabase::getTable() . '.id AS dbid',
+            GDatabase::getTable() . '.name',
+            DatabaseInstance::getTable() . '.id AS iid',
+            DatabaseInstance::getTable() . '.name AS instance_name',
+            'port',
```
Comment (trasher) [inline inc/inventory/asset/database.class.php:78]:
```suggestion
            'instance_port',
```

File: inc/define.php (hunk 10 above)
Comment (trasher) [reply inc/define.php:583]: ping @orthagh @AdrienClairembault

File: inc/inventory/asset/database.class.php (hunk 25 above)
Comment (trasher) [reply inc/inventory/asset/database.class.php:119]: Yes, I'm still not sure how to handle updates properly... :)

File: inc/define.php (hunk 10 above)
Comment (AdrienClairembault) [reply inc/define.php:583]: Our source for icons is https://www.flaticon.com/authors/smashicons (it's in pics/impact/README.txt if you forget the url).

For a database, the "classic" icon seem to be this one : https://www.flaticon.com/free-icon/database_148825

**Review by orthagh — APPROVED** [review]

**Pull request merged**